### PR TITLE
Switching from poetry to uv 

### PR DESCRIPTION
**Pull request by apchytr**
**Context:** [astral-sh/uv ](https://github.com/astral-sh/uv) is an alternative package manager written in rust. The aim here is to replace poetry with uv. 

**Description of the Change:** Updating references to poetry with uv. Updating pyproject.toml and the lock file. 

**Benefits:** uv is faster than poetry. Additionally, our pyproject.toml file is now up to the latest PEP standards. 

**Possible Drawbacks:** There may be some features in poetry that are currently missing in uv. 

Note: changes will need to be propagated to Mr Kite

### CODE REVIEW

**Comment by ziofil** [comment]
zippity fast!

**Review by ziofil — APPROVED** [review]
thanks! this is awesome

**Review by timmysilv — COMMENTED** [review]
is the lockfile from scratch, as in, packages are now bumped to latest instead of whatever was in `poetry.lock`?

File: pyproject.toml [hunk 1]
```diff
@@ -24,61 +30,56 @@ classifiers = [
     "Programming Language :: Python :: 3 :: Only",
     "Topic :: Scientific/Engineering :: Physics",
 ]
+requires-python = ">=3.9,<3.13"
+dependencies = [
+    "grpcio == 1.60.0",
+    "numpy >=1.23.5, <2",
```
[the remark this answers is not in the document]
Comment (apchytr) [reply pyproject.toml:36]: Doesn't seem like it. When looking at [PEP 440](https://peps.python.org/pep-0440/) it's not mentioned so maybe it isn't considered a PEP standard?

File: Makefile [hunk 2]
```diff
@@ -27,19 +27,19 @@ install:
...
+	uv sync
 
 
 .PHONY: install-all
 install-all:
 ifndef PYTHON3
 	@echo "To install Mr Mustard you need to have Python 3 installed"
 endif
-	poetry install --all-extras --with dev,doc
+	uv sync --all-extras --group doc
```
[the remark this answers is not in the document]
Comment (apchytr) [reply Makefile:38]: Nope! It's just included by default

**Comment by apchytr** [comment]
> is the lockfile from scratch, as in, packages are now bumped to latest instead of whatever was in `poetry.lock`?

It was generated from scratch yes. As to whether this is an issue I'm not sure. What do you think @timmysilv?

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pyproject.toml:36]: i hate it anyway so that's fine

**Review by timmysilv — APPROVED** [review]
i don't think it's an issue, but it's something of note! I'll list the changes here for posterity in case it causes issues in the future.

<details>
<summary>Changed packages</summary>
Changed aiosignal from 1.3.1 to 1.3.2</br>
Changed astroid from 3.3.5 to 3.3.8</br>
Changed asttokens from 2.4.1 to 3.0.0</br>
Changed attrs from 24.2.0 to 24.3.0</br>
Changed black from 24.10.0 to 22.12.0</br>
Changed certifi from 2024.8.30 to 2024.12.14</br>
Changed charset-normalizer from 3.4.0 to 3.4.1</br>
Changed click from 8.1.7 to 8.1.8</br>
Changed cloudpickle from 3.1.0 to 3.1.1</br>
Changed contourpy from 1.3.0 to 1.3.1</br>
Changed coverage from 7.6.8 to 7.6.10</br>
Changed dask from 2024.8.0 to 2025.1.0</br>
Changed docutils from 0.20.1 to 0.21.2</br>
Changed fastjsonschema from 2.21.0 to 2.21.1</br>
Changed flatbuffers from 24.3.25 to 24.12.23</br>
Changed fonttools from 4.55.0 to 4.55.3</br>
Changed fsspec from 2024.10.0 to 2024.12.0</br>
Changed identify from 2.6.3 to 2.6.5</br>
Changed ipython from 8.18.1 to 8.31.0</br>
Changed jinja2 from 3.1.4 to 3.1.5</br>
Changed keras from 3.7.0 to 3.8.0</br>
Changed kiwisolver from 1.4.7 to 1.4.8</br>
Changed llvmlite from 0.42.0 to 0.43.0</br>
Changed matplotlib from 3.9.2 to 3.10.0</br>
Changed networkx from 3.2.1 to 3.4.2</br>
Changed numba from 0.59.1 to 0.60.0</br>
Changed optree from 0.13.1 to 0.14.0</br>
Changed pillow from 11.0.0 to 11.1.0</br>
Changed protobuf from 5.29.0 to 5.29.3</br>
Changed pyaml from 24.9.0 to 25.1.0</br>
Changed pygments from 2.18.0 to 2.19.1</br>
Changed pylint from 3.3.1 to 3.3.3</br>
Changed pyparsing from 3.2.0 to 3.2.1</br>
Changed ray from 2.39.0 to 2.40.0</br>
Changed referencing from 0.35.1 to 0.36.1</br>
Changed rpds-py from 0.21.0 to 0.22.3</br>
Changed scikit-learn from 1.5.2 to 1.6.1</br>
Changed scikit-optimize from 0.9.0 to 0.10.2</br>
Changed scipy from 1.13.1 to 1.15.1</br>
Changed six from 1.16.0 to 1.17.0</br>
Changed sphinx-automodapi from 0.15.0 to 0.18.0</br>
Changed urllib3 from 2.2.3 to 2.3.0</br>
Changed virtualenv from 20.28.0 to 20.29.1</br>
Changed wrapt from 1.17.0 to 1.17.2</br>
</details>
